[GRO-194] Documenting role-based credentials (#73)
hashboard-hq/documentation: pages/docs/database-connections/role-based-credentials.mdx

---
description: Role-based data credentials allow you to control what users can access from the same data connection.
---

import { Callout } from "nextra-theme-docs";

# Role-Based Database Credentials

All Hashboard data connections can be optionally configured to use additional credentials which can be assigned to users on a per-role basis. 

After setting up database-level permissions within the data warehouse you use with Hashboard, these credentials can be added within Hashboard and then assigned to your Hashboard project's roles.

> 

Reviewer: I think we can do without this sentence? Seems essentially equivalent to what's above.

> This enables projects to control what data users can access when viewing models, saved explorations, and dashboards built from the same data connection.

Reviewer 2: This reads a bit confusing to me - are you saying that this will enable the specific roles in projects? Right now it reads that there will be multiple projects.

Reviewer: +1, should instead read "This enables you to control..."

Role-based credentials can be used to restrict access to certain tables, or even certain rows or columns within a table, based on a user's role.

### Adding credentials

The first step to adding role-based credentials is to create a [database connection](/docs/database-connections).

Note that the credentials specified in the top-level of the database connection form is the *default* credential. 
That is, unless specified otherwise, all users will connect via the information provided in the database connection form.

<Callout>
    If a user doesn't have any specified credentials within their roles, they will use the default credential for the data connection.

>     Therefore, it's best practice for the default credential to be the most restrictive credential permissions-wise.

Reviewer: While this does seem like a safer recommendation, I think the workflow _could_ end up being the opposite in some cases - it might depend on the company's specific use case. For example: if only a few people can't see certain columns in a db, it might be simpler to make the default the less restrictive permission and add individuals who need restrictions on an ad hoc basis.

I'll defer approval on what we do here to @vndrewlee on the product side. Could see going with this, or could also see describing something more nuanced.

Author: I definitely think that different scenarios could call for different permissioning structures. Also I forgot to mention public links here.

Maybe I can change this to something that still calls this out but is less opinionated, like 

"Make sure your users are seeing the data you intend them to: if a user doesn't have any specified credentials within their roles, they will use the default credential for the data connection. Default credentials are also used in public links, if they are enabled." 

interested in hearing what people think abt this though! Could be something we touch on in a blog post as well

Reviewer: As per our convo about what we're going with here (the way you suggested!) - let's make sure to make this opinion clearly stated now! Perhaps even more opinionated than it is now, such as:

`Hashboard assumes that your default credential is the most restrictive. etc. etc.`

Author: Thanks for pointing this out– I made this warning stronger

"If a user doesn't have any specified credentials within their roles, they will use the default credential for the data connection. Default credentials are also used in public links, if they are enabled.

Therefore, Hashboard assumes your default credential is the most restrictive credential permissions-wise. Make sure your default credential doesn't allow access to sensitive data that you would not want accessible to all users in your organization."

</Callout>

At the bottom of the data connection form, you'll see a section for **Role-based credentials**.

Click **Add credential** and fill out the fields provided. Give the credential a unique and descriptive name, 
as it will be used to assign to roles later on.

Once you're done, make sure to **Test** the credentials before saving to ensure they are connecting properly.

Finally, click **Save**.

> #### Settings

Reviewer: I would rename this to `Credential fields` and reference it in the `Click **Add credential** and fill out...` step above (e.g. see `Credential fields` below for more details on ...`).

Alternatively, you could just include it under that bullet.

The settings for a role-based credential depend on which data warehouse you're connecting to.

* [**Snowflake**](/docs/database-connections/snowflake): User, password, and role are required.

* [**BigQuery**](/docs/database-connections/bigquery): JSON key is required.

* [**PostgreSQL/Redshift**](/docs/database-connections/postgresql): User and password are required.

* [**MySQL**](/docs/database-connections/mysql): User and password are required.

* [**Athena**](/docs/database-connections/athena): AWS Access Key ID and AWS Secret Access Key are required.

* [**ClickHouse**](/docs/database-connections/clickhouse): User and password are required.

* [**MotherDuck**](/docs/database-connections/motherduck): Service token is required.

### Assigning credentials to roles

> For information about how to set up and assign roles, see [Users and Permissions](/docs/project-management/users-and-permissions).

Reviewer: Same nit here about numbering these

---
description: Role-based data credentials allow you to control what users can access from the same data connection.
---

import { Callout } from "nextra-theme-docs";

# Role-Based Database Credentials

All Hashboard data connections can be optionally configured to use additional credentials which can be assigned to users on a per-role basis. 

This enables you to control what data Hashboard users can access when viewing models, saved explorations, and dashboards built from the same data connection.

> Role-based credentials can be used to restrict access to certain tables, or even certain rows or columns within a table, based on a user's role.

Reviewer 3: suggestion: make this the first sentence

### Adding credentials

1. First, create a [database connection](/docs/database-connections). The credential specified in the top-level of the database connection form is the *default* credential. 

<Callout>
    If a user doesn't have any specified credentials within their roles, they will use the default credential for the data connection.

>     Default credentials are also used in public links, if they are enabled.

Reviewer 3: ```suggestion
    Default credentials are also used in public links.
```

>     Therefore, Hashboard assumes your default credential is the most restrictive credential permissions-wise.

Reviewer 3: ```suggestion
    We recommend making your default credential the most restricted credential.
```
